File: java/tencent-new/arrayandstr/RemoveDuplicates.java
package arrayandstr;

import java.util.Arrays;

/**
 * @author majexh
 */
public class RemoveDuplicates {

    public int removeDuplicates(int[] nums) {
        if (nums.length == 0) return 0;
        int last = 0, fast = 1;
        while (fast < nums.length) {
            if (nums[last] != nums[fast]) {
                // 找到第一个不等的了
                nums[++last] = nums[fast];
            }
            fast++;
        }
        System.out.println(Arrays.toString(nums));
        return last + 1;
    }

    public static void main(String[] args) {
        System.out.println(new RemoveDuplicates().removeDuplicates(new int[]{ 0,0,1,1,1,2,2,3,3,4 }));
    }
}

File: go/tencent/array_and_str/remove_duplicates.go
package main

import "fmt"

func main() {
	a := []int{0, 0, 1, 1, 1, 2, 2, 3, 3, 4}
	fmt.Println(removeDuplicates(a), a)
}

func removeDuplicates(nums []int) int {
	fast, last := 0, 0

	for fast < len(nums) {
		for fast < len(nums) && nums[fast] == nums[last] {
			fast++
		}
		if fast == len(nums) {
			break
		}
		last++
		nums[last] = nums[fast]
	}

	return last + 1
}
